plugin/templates/multi_pattern_resource_name.mustache

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package {{package}};
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.api.core.BetaApi;
+import com.google.api.pathtemplate.PathTemplate;
+import com.google.api.pathtemplate.ValidationException;
+import {{resource_name_global_package_name}}.ResourceName;
+import java.util.Map;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * AUTO-GENERATED DOCUMENTATION AND CLASS
+ */
+@javax.annotation.Generated("by GAPIC protoc plugin")
+public class {{class_name}} implements ResourceName {
+  
+  @Deprecated
+  protected {{class_name}}() { }
+
+  {{#patterns}}
+  {{#is_formattable}}
+  private static final PathTemplate {{upper_underscore}}_PATH_TEMPLATE =
+      PathTemplate.createWithoutUrlEncoding("{{pattern_string}}");
+  {{/is_formattable}}
+  {{#is_fixed}}
+  private static final String {{upper_underscore}}_FIXED_VALUE =
+  		"{{pattern_string}}";
+  private static final {{class_name}} {{upper_underscore}}_INSTANCE =
+  		new {{class_name}}("{{pattern_string}}");
+  {{/is_fixed}}
+  {{/patterns}}
+
+  private volatile Map<String, String> fieldValuesMap;
+  private PathTemplate pathTemplate;
+  private String fixedValue;
+
+  {{#format_fields}}
+  private String {{lower_camel_symbol}};
```

Comment:
The resource name classes do not expose setters - which make these fields nominally final.

The reason I did not make them final is that a certain field may not be used by all resource name patterns - in which case the field may be `null`. If they are final, in the constructors we will need to explicitly do
```java
private BookName(Builder builder) {
  project = builder.project;
  book = builder.book;
  location = null;
}
```
Otherwise JDK will complain that the final field `location` is not initialized.